doc/developer/design/20240919_dataflow_expiration.md

```diff
@@ -0,0 +1,131 @@
+
+## The Problem
+
+Temporal filters currently require Materialize to maintain all future retractions
+of data that is currently visible. For long windows, the retractions could be at
+timestamps beyond our next scheduled restart, which is typically our weekly
+DB release.
+
+For instance, in the example below, the temporal filter in the `last_30_days`
+view causes two diffs to be generated for every row inserted into `events`, the
+row itself and a retraction 30 days later. However, if the replica is
+restarted in the next few days, the retraction diff is never processed, making
+it redundant to keep the diff waiting to be processed.
+
+```sql
+-- Create a table of timestamped events.
+CREATE TABLE events (
+  content TEXT,
+  event_ts TIMESTAMP
+);
+-- Create a view of events from the last 30 seconds.
+CREATE VIEW last_30_days AS
+SELECT event_ts, content
+FROM events
+WHERE mz_now() <= event_ts + INTERVAL '30 days';
+
+INSERT INTO events VALUES ('hello', now());
+
+COPY (SUBSCRIBE (SELECT event_ts, content FROM last_30_days)) TO STDOUT;
+1727130590201   1       2023-09-23 22:29:50.201 hello  -- now()
+1729722590222   -1      2023-10-23 22:29:50.222 hello  -- now() + 30 days
+```
+
+Dataflows with large temporal windows (e.g. a year) can generate a large number
+of retractions that consume memory and CPU but are never used. Instead, all
+such retractions can be dropped.
+
+## Success Criteria
+
+When temporal filters are in use, retraction diffs associated with timestamps
+beyond a set expiration time can be dropped without affecting correctness,
+resulting in lower memory and CPU utilization from halving the number of
+processed diffs.
+
+## Solution Proposal
+
+A new LaunchDarkly feature flag is introduced that specifies an _expiration
+offset_ (a `Duration`). The _replica expiration_ time is computed as the offset
+added to the start time of the replica. Dataflows matching certain
+criteria (detailed below) are then configured with a _dataflow expiration_
+derived from the replica expiration. Diffs generated in these dataflows beyond
+the dataflow expiration are dropped. To ensure correctness, panic checks are
+added to these dataflows that ensure that the frontier does not exceed the
+dataflow expiration before the replica is restarted.
+
+An overview of the logic used for these features is as follows:
+```
+# Consider the `upper` for different dataflows
+if mv_view_with_constant_values:
+  upper := []
+else if mv_with_refresh_every:
+  upper := [next_refresh()]
+else:
+  upper := [write_frontier()]
+
+# The `upper` for a dataflow considering all its transitive inputs
+inputs_upper := meet(for all inputs i: i_upper)
+
+# Dataflow expiration logic
+if compute_replica_expiration_offset is not set:
+  dataflow_replication := []
+else for dataflows of type in [materialized view, index, subscribe]:
+  replica_expiration := replica_start + compute_replica_expiration_offset
+  if dataflow_timeline is not EpochMilliseconds:
```

Comment:
Add a note that materialized views and indexes that do not depend on any source or table will not be in the epoch milliseconds timeline.